pkg/controller/daemon/util/daemonset_util.go

```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"strconv"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+	extensions "k8s.io/kubernetes/pkg/apis/extensions/v1beta1"
+	labelsutil "k8s.io/kubernetes/pkg/util/labels"
+)
+
+func GetPodTemplateWithHash(ds *extensions.DaemonSet, dsPodTemplateSpecHash uint32) v1.PodTemplateSpec {
+	obj, _ := api.Scheme.DeepCopy(ds.Spec.Template)
+	dsPodTemplateSpecHashStr := strconv.FormatUint(uint64(dsPodTemplateSpecHash), 10)
+	template := obj.(v1.PodTemplateSpec)
+	template.ObjectMeta.Labels = labelsutil.CloneAndAddLabel(
+		ds.Spec.Template.ObjectMeta.Labels,
+		extensions.DefaultDaemonSetUniqueLabelKey,
+		dsPodTemplateSpecHashStr,
+	)
+	return template
+}
+
+func IsPodUpdated(dsPodTemplateSpecHash uint32, pod *v1.Pod) bool {
+	curPodTemplateSpecHash, hashExists := pod.ObjectMeta.Labels[extensions.DefaultDaemonSetUniqueLabelKey]
+	dsPodTemplateSpecHashStr := strconv.FormatUint(uint64(dsPodTemplateSpecHash), 10)
+	if hashExists {
+		if curPodTemplateSpecHash == dsPodTemplateSpecHashStr {
+			return true
+		}
+	} else {
+		// XXX: Hash does not exist. It's old pod. For now returning true.
```

Comment:
Also about new pods with the hash, afaik these are not immutable so something may change in their spec and the ds controller will continue thinking they are new, right? This is part of a bigger problem, @lukaszo can you open a separate issue about it? I believe we should validate changes in resources owned by other resources but it needs more discussion.